fix(components/modals): hide non modal elements from screen readers

Blackbaud-SandhyaRajasabeson opened the pull request:
add aria-hidden to siblings of the current modal and modal host so that screen readers cannot
navigate outside the modal while it is open

blackbaud-sky-build-user commented:
[Storybook preview](https://blackbaud.github.io/skyux-pr-preview/397/storybook/)

Component Storybooks:

- [ag-grid-storybook](https://blackbaud.github.io/skyux-pr-preview/397/storybooks/ag-grid-storybook/)
- [modals-storybook](https://blackbaud.github.io/skyux-pr-preview/397/storybooks/modals-storybook/)
- [pages-storybook](https://blackbaud.github.io/skyux-pr-preview/397/storybooks/pages-storybook/)
- [tabs-storybook](https://blackbaud.github.io/skyux-pr-preview/397/storybooks/tabs-storybook/)

Apps:

- [code-examples](https://blackbaud.github.io/skyux-pr-preview/397/code-examples/)
- [integration](https://blackbaud.github.io/skyux-pr-preview/397/integration/)
- [playground](https://blackbaud.github.io/skyux-pr-preview/397/playground/)

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -14,6 +15,7 @@ import {
 } from '@skyux/core';
 
 import { takeWhile } from 'rxjs/operators';
+import { __values } from 'tslib';
```
Eh?

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal.component.ts:
```diff
@@ -25,7 +25,7 @@ import { SkyModalScrollShadowEventArgs } from './modal-scroll-shadow-event-args'
 
 let skyModalUniqueIdentifier = 0;
-
+//get modal host siblings and put either aria hidden or display hidden in css
```
This comment seems to be for nothing. Am I missing something?

Blackbaud-SteveBrush replied:
Should this comment be here?

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -109,15 +117,89 @@ export class SkyModalHostComponent implements OnDestroy {
       injector
     );
 
+    //const modalsAfter = this.hostElement.nativeElement.children;
```
Extra commented out line?

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -109,15 +117,89 @@ export class SkyModalHostComponent implements OnDestroy {
     });
 
+    const modalsBefore = Array.from(
+      this.hostElement.nativeElement.parentElement.children
```
Is there a reason that we can't  just use `this.target.element.nativeElement` here? It feels odd that we are adding another element just for grabbing the parent element when we already have another sibling element.

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -109,15 +117,89 @@ export class SkyModalHostComponent implements OnDestroy {
+    // hidding all elements at the modal-host level from screenreaders
+    const ariaPreviousValueSiblings = new Map<Element, string | null>();
+    const siblings =
+      document.querySelector('sky-modal-host').parentElement.children;
```
Any logic that is grabbing things off the actual DOM API should go into an adapter service. So in this case they should go into the `SkyModalAdapterService`.

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -109,15 +117,89 @@ export class SkyModalHostComponent implements OnDestroy {
+      if (
+        sibling !== document.querySelector('sky-modal-host') &&
+        !sibling.hasAttribute('aria-live') &&
+        sibling.nodeName !== 'SCRIPT' &&
```
Should we do a `sibling.nodeName.toLowerCase()` just to be safe and then use 'script'? That way if some browser decides to return 'script' we will handle it.

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -109,15 +117,89 @@ export class SkyModalHostComponent implements OnDestroy {
+
+    // hiding all sibling modals from screenreaders
+    const ariaPreviousValueChildren = new Map<Element, string | null>();
+    const children = document.querySelector('sky-modal-host').children;
```
Nitpicky - could we call this `hostChildren` and then the later variable `hostChild` just to make it clearer what children we have?

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -109,15 +117,89 @@ export class SkyModalHostComponent implements OnDestroy {
 
+    // hidding all elements at the modal-host level from screenreaders
+    const ariaPreviousValueSiblings = new Map<Element, string | null>();
+    const siblings =
```
Nitpicky again - could we call this `hostSiblings`? It took me a minute to figure out what we were grabbing here.

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -109,15 +117,89 @@ export class SkyModalHostComponent implements OnDestroy {
+      const sibling = siblings[i];
+
+      if (
+        sibling !== document.querySelector('sky-modal-host') &&
```
Can we inject `ElementRef` into the constructor of this? Then we could do `sibling !== this.elRef.nativeElement` and don't have to query the DOM again.

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -109,15 +117,89 @@ export class SkyModalHostComponent implements OnDestroy {
+
+      if (
+        sibling !== document.querySelector('sky-modal-host') &&
+        !sibling.hasAttribute('aria-live') &&
```
For my own knowledge - can you explain why we aren't placing `aria-hidden` on elements which have `aria-live`?

Blackbaud-SandhyaRajasabeson replied:
i was basing this off of how angular handles `aria-hidden` with their modals. if we put aria-hidden on aria-live elements they won't announce and won't announce even after the modal is removed so it would be information that never gets to the user.

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -109,15 +117,89 @@ export class SkyModalHostComponent implements OnDestroy {
+    );
+
+    // getting the element ref to the modal that was just opened
+    const modalElement = modalsAfter.filter(
```
It seems like the only reason we are querying the modals before and after are to get the new opened item. However, we should be able to get that by using `modalComponentRef.location.nativeElement`

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -109,15 +117,89 @@ export class SkyModalHostComponent implements OnDestroy {
+
     function closeModal() {
+      // unhiding modal-host siblings
+      ariaPreviousValueSiblings.forEach((previousValue, element) => {
```
What happens if a lower modal closes before the most upper modal (not a common case - but could happen via the typescript service technically.

Blackbaud-TrevorBurch replied:
I think that the `children` would most likely be ok as the lower modal wouldn't have the new `children` and so it probably would be fine. However, if I'm reading this correctly - closing the lowest modal would reenable all siblings even though an upper modal is still open.

Blackbaud-SandhyaRajasabeson replied:
im going to add a unit test that opens two modals and then closes the first one just to confirm that that is what would happen. but if that is so, im thinking make the list of aria hidden modals into a global variable and check that when any modal is closed.

Blackbaud-TrevorBurch reviewed (changes requested):
@Blackbaud-SandhyaRajasabeson - some comments for you.

Commit:
fix(components/modals): hide non modal elements from screen readers

add aria-hidden to siblings of the current modal and modal host so that screen readers cannot
navigate outside the modal while it is open

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-adapter.service.ts:
```diff
@@ -47,4 +47,36 @@ export class SkyModalAdapterService {
+        element.nodeName.toLowerCase() !== 'script' &&
+        element.nodeName.toLowerCase() !== 'style'
+      ) {
+        hiddenElements.set(element, element.getAttribute('aria-hidden'));
```
Mind putting a comment above this line to explain why we're saving the existing `aria-hidden` value? I know why (because we talked about it during Flex) but it would be good to mention it here for posterity.

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-adapter.service.ts:
```diff
@@ -47,4 +47,36 @@ export class SkyModalAdapterService {
+
+  public removeAriaHidden(hiddenElements: Map<Element, string | null>): void {
+    hiddenElements.forEach((previousValue, element) => {
+      if (previousValue) {
```
I would add a comment above this line explaining why there's a condition. Something to the effect of 
```
// Restore the initial value of `aria-hidden` if applicable.
```

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -139,11 +155,38 @@ export class SkyModalHostComponent implements OnDestroy {
+
+    if (SkyModalHostService.openModalCount == 1) {
+      this.hideModalHostSiblings();
+      ariaPreviousValueSiblings = this.#ariaPreviousValueSiblings;
```
What's the value in setting this again? Isn't it getting set up above? 

More importantly, I believe you're creating a local copy of the `#ariaPreviousValueSiblings` class property because you can't use `this` in the `closeModal` function. I would instead convert the `closeModal` function into a "fat arrow" function and just use the `this.#ariaPreviousValueSiblings` property directly, so that we can avoid adding needless "shadow" variables.

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -139,11 +155,38 @@ export class SkyModalHostComponent implements OnDestroy {
+    const hostElement = this.#elRef.nativeElement;
+    let allModals = hostElement.children;
+
+    if (SkyModalHostService.openModalCount == 1) {
```
I could be wrong, but why would we check if there's only one modal opened before hiding all siblings? What happens if an element is added to the BODY tag when a second modal has been opened? Shouldn't we hide any newly created siblings when that happens?

Blackbaud-SteveBrush replied:
(I would add a unit test that predicts that exact scenario to confirm.)

Blackbaud-SandhyaRajasabeson replied:
my thinking was that since the user is in the modal there wouldn't be things added behind it so if we just hide things the first time a modal is opened we would be good. but i guess that's idealic thinking.

Blackbaud-SteveBrush replied:
Yeah, we can't control what other scripts are doing on the page. Unless I'm missing something, it wouldn't hurt to hide all siblings every time a new modal opens. Would be good to provide a test for this scenario as well.

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal.service.spec.ts:
```diff
@@ -176,9 +176,137 @@ describe('Modal service', () => {
+    div.remove();
+  }));
+
+  it('what does it do with layered modals', fakeAsync(() => {
```
All `it` blocks should start with the word "should". You also have two tests named the same thing.

Blackbaud-TrevorBurch replied:
We tend to write our spec descriptions in a `should...` format that describes what should happen. I don't know based on the description here or in the next test what they "should" be doing.

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal.service.spec.ts:
```diff
@@ -176,9 +176,137 @@ describe('Modal service', () => {
+    div.remove();
+  }));
+
+  it('should keep hidden status elements at the host level consistent', fakeAsync(() => {
```
For clarity, I would reword the test to something like, "should restore sibling elements' initial 'aria-hidden' values after all modals are closed".

Blackbaud-SandhyaRajasabeson replied:
do we not want to have the "should" statements be generic with no "code" language?

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal.service.spec.ts:
```diff
@@ -176,9 +176,137 @@ describe('Modal service', () => {
+    closeModal(modal);
+  }));
+
+  it('should not hide modal host from screen readers when opening a modal', fakeAsync(() => {
```
It might be beneficial to wrap these tests in their own `describe` block so they're easier to digest.

Blackbaud-SteveBrush replied:
Does this `it` need to be in the `describe` block below?

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal.service.spec.ts:
```diff
@@ -176,9 +176,137 @@ describe('Modal service', () => {
+
+    const lowerModal = openModal(ModalTestComponent, { fullPage: false });
+    const list = host.children;
+    list[3].id = 'lowerModal';
```
I was initially very confused that the host had more than 2 children (since you only opened two modals in this test). Once I looked in the host's template I could see that we have a few child elements baked-in. That being said, simply reading the host's children can be a bit brittle (what happens when we decide to add another element to the host's template?). Instead, I would create a helper function `getModalEls()` that would return a `NodeListOf<Element>` for all modal elements so that future contributors to this test are not confused.
```
function getModalEls() {
  return document.querySelectorAll('sky-test-cmp');
}
```
Then, in your tests, you can do the following:
```
const modalEls = getModalEls();
modalEls.item(0).blah
```

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal.service.spec.ts:
```diff
@@ -176,9 +176,137 @@ describe('Modal service', () => {
+
+  it('what does it do with layered modals', fakeAsync(() => {
+    openModal(ModalTestComponent, { fullPage: false });
+    document.querySelector('sky-test-cmp').id = 'sibling';
```
Can we add more descriptive IDs than "sibling" and "lowerModal"? I was having trouble tracking the logic of the test with those names. Also, the "lowerModal" is opened after the "sibling" modal; shouldn't it be the "upperModal" instead?

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal.service.spec.ts:
```diff
@@ -176,9 +176,137 @@ describe('Modal service', () => {
+
+  it('should hide and unhide modal siblings from screen readers', fakeAsync(() => {
+    openModal(ModalTestComponent, { fullPage: false });
+    document.querySelector('sky-test-cmp').id = 'sibling';
```
Any place where you're querying for modals, you can use the new `getModalEls()` function (see above).

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -139,11 +155,38 @@ export class SkyModalHostComponent implements OnDestroy {
+      allModals[allModals.length - 1] == modalElement
+    ) {
+      allModals[allModals.length - 2].setAttribute('aria-hidden', true);
+    }
```
I would use either `nextElementSibling` or `previousElementSibling` for this. https://developer.mozilla.org/en-US/docs/Web/API/Element/previousElementSibling

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -139,11 +155,38 @@ export class SkyModalHostComponent implements OnDestroy {
+      } else if (SkyModalHostService.topModal == hostService) {
+        // if there are more than 1 modal then unhide the one behind this one before closing it
+        allModals = hostElement.children;
+        allModals[allModals.length - 2].removeAttribute('aria-hidden');
```
Same here, RE: `nextElementSibling` or `previousElementSibling`.

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -139,11 +155,38 @@ export class SkyModalHostComponent implements OnDestroy {
+      // unhide siblings if last modal is closing
+      if (SkyModalHostService.openModalCount == 1) {
+        adapter.removeAriaHidden(ariaPreviousValueSiblings);
+        ariaPreviousValueSiblings.clear();
```
This feels like a wholistic action, so I would put this in the host's `ngOnDestroy()` method.

Blackbaud-SandhyaRajasabeson replied:
that would only be called when `sky-modal-host` is destroyed right? `sky-modal-host` isn't removed from the dom even when all the modals all closed so wouldn't it make more sense for this to be called when the last modal is closed and not when host is destroyed

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -57,8 +58,10 @@ export class SkyModalHostComponent implements OnDestroy {
   #modalHostContext: SkyModalHostContext;
+  #elRef: ElementRef;
+
+  #ariaPreviousValueSiblings = new Map<Element, string | null>();
```
Since the `SkyModalAdapterService` is provided by this component, its state will be tied to this component as well. Meaning, the adapter service could manage all of the sibling elements itself. This would also keep the modal component from being "polluted" with DOM features and types (the adapter service is designed to abstract away the DOM-specific aspects of our component class).

If possible, I would move the `#ariaPreviousValueSiblings` property to the adapter service and have it manage the `aria-hidden` states and elements itself. The only thing you should be calling in this component would be:
```
this.#adapter.hideHostSiblingsForScreenReaders(this.#elRef);
this.#adapter.showHostSiblingsForScreenReaders(this.#elRef);

this.#adapter.hidePreviousModalForScreenReaders();
this.#adapter.showPreviousModalForScreenReaders();
```

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-adapter.service.ts:
```diff
@@ -47,4 +47,36 @@ export class SkyModalAdapterService {
     this.#bodyEl.classList.remove(className);
   }
+
+  public addAriaHidden(
```
Nitpicky - could we add a comment here just stating that we return back the initial value of the elements' `aria-hidden` attribute?

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-adapter.service.ts:
```diff
@@ -47,4 +47,36 @@ export class SkyModalAdapterService {
+    elementsToHide: HTMLCollection,
+    self: Element | unknown
+  ): Map<Element, string | null> {
+    const hiddenElements = new Map<Element, string | null>();
```
Would a better name for this be `elementsPreviousValue` or something similar since this map isn't just the elements but also is that.

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-adapter.service.ts:
```diff
@@ -47,4 +47,36 @@ export class SkyModalAdapterService {
+    return hiddenElements;
+  }
+
+  public removeAriaHidden(hiddenElements: Map<Element, string | null>): void {
```
I wonder if we should call this `restoreOrRemoveAriaHidden` and then name the parameter something similar to whatever we name the variable in the field above (see my comment there).

We aren't always removing it in the case where it previously was set.

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -104,6 +119,7 @@ export class SkyModalHostComponent implements OnDestroy {
 
     const adapter = this.#adapter;
     const modalOpener: HTMLElement = adapter.getModalOpener();
+    let ariaPreviousValueSiblings = this.#ariaPreviousValueSiblings;
```
I don't follow why we need this local variable? It seems like all that is happening is that it is set to the class level field and then cleared at the end. Couldn't we just reference the class level field all the time?

Blackbaud-TrevorBurch commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -139,11 +155,38 @@ export class SkyModalHostComponent implements OnDestroy {
+      } else if (SkyModalHostService.topModal == hostService) {
+        // if there are more than 1 modal then unhide the one behind this one before closing it
+        allModals = hostElement.children;
+        allModals[allModals.length - 2].removeAttribute('aria-hidden');
```
What if there are three modals and the middle once closes. Wouldn't this unhide the 1st and leave both the 1st and 3rd unhidden?

Blackbaud-SandhyaRajasabeson replied:
nope! it checks that the top modal is the one that is closing and only then unhides any modals

Blackbaud-SandhyaRajasabeson replied:
there is a unit test for this scenario also just to make sure no changes break that expectation https://github.com/blackbaud/skyux/blob/modal-navigation-accessibility/libs/components/modals/src/lib/modules/modal/modal.service.spec.ts#L239

Commit:
add comments on aria hidden and convert to fat arrow function

Commit:
comments on unit test

Commit:
close every modal that's opened in the unit test

Commit:
use new function

Commit:
using previousElementSibling and topModal

Commit:
moving dom checking to adapter

Commit:
function names

Commit:
move accesibiliyt unit test into its own block

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -139,11 +142,34 @@ export class SkyModalHostComponent implements OnDestroy {
+
+    const closeModal = () => {
+      // unhide siblings if last modal is closing
+      if (SkyModalHostService.openModalCount == 1) {
```
Needs a triple-equals `===`.

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -139,11 +142,34 @@ export class SkyModalHostComponent implements OnDestroy {
+      // unhide siblings if last modal is closing
+      if (SkyModalHostService.openModalCount == 1) {
+        this.#adapter.unhideOrRestoreHostSiblings();
+      } else if (SkyModalHostService.topModal == hostService) {
```
Needs a triple-equals `===`.

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-adapter.service.ts:
```diff
@@ -47,4 +48,50 @@ export class SkyModalAdapterService {
     this.#bodyEl.classList.remove(className);
   }
+
+  public hideHostSiblings(hostElRef: ElementRef): void {
```
For posterity, let's put a JSDocs comment above this method describing what it does.

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-adapter.service.ts:
```diff
@@ -47,4 +48,50 @@ export class SkyModalAdapterService {
+    }
+  }
+
+  public unhideOrRestoreHostSiblings(): void {
```
For posterity, let's put a JSDocs comment above this method describing what it does.

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-adapter.service.ts:
```diff
@@ -13,6 +13,7 @@ export class SkyModalAdapterService {
   #bodyEl: HTMLElement;
 
   #windowRef: SkyAppWindowRef;
+  #hostSiblingPreviousValue = new Map<Element, string | null>();
```
This property name is ambiguous. Got a better name that explains what kinds of values it's tracking?

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-adapter.service.ts:
```diff
@@ -47,4 +48,50 @@ export class SkyModalAdapterService {
+    this.#hostSiblingPreviousValue.clear();
+  }
+
+  public hidePreviousModal(topModal: Element): void {
```
For all four of these methods, the word "hide" is misleading since the modal is not being hidden for sighted users. Even though it's a tad wordy, I would add the qualifier "FromScreenReaders" to the method names so it's 100% clear.
```
public hidePreviousModalFromScreenReaders
```

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-adapter.service.ts:
```diff
@@ -47,4 +48,50 @@ export class SkyModalAdapterService {
+    }
+  }
+
+  public unhidePreviousModal(topModal: Element): void {
```
Adapter services should never receive an `Element` but rather an `ElementRef`. The idea behind this decision is that these adapter services can be switched out with other platform adapters (besides the DOM). For example, if we wanted to use the modal stuff in a mobile app, we could switch the DOM adapter with a mobile phone adapter. Since the mobile phone adapter wouldn't know what an `Element` is, we should always pass the `ElementRef` since that will represent either a DOM element or a mobile phone "element".

This should be updated in a few places above, too.

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -139,11 +142,34 @@ export class SkyModalHostComponent implements OnDestroy {
     this.#registerModalInstance(modalInstance);
 
-    function closeModal() {
+    if (SkyModalHostService.openModalCount == 1) {
```
Needs a triple-equals `===`.

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-host.component.ts:
```diff
@@ -139,11 +142,34 @@ export class SkyModalHostComponent implements OnDestroy {
+    }
+    if (
+      SkyModalHostService.openModalCount > 1 &&
+      SkyModalHostService.topModal == hostService
```
Needs a triple-equals `===`.

Commit:
move one more unit test

Commit:
unit test

Commit:
function names, docs, more unit tests

Commit:
top modal to element ref

Blackbaud-SandhyaRajasabeson committed:
Merge branch 'main' into modal-navigation-accessibility
